### ISSUE TITLE
Allowlist and blocklist for detector path lists

### ISSUE THREAD

**Issue by landauermax** [ait-aecid/logdata-anomaly-miner#266]
allowlisted_paths in ECD should be named blocklisted_paths, since these paths are not considered for detection.

allowlisted_paths should also exist, but does the oppsite: Only when one of the paths in the logatom match dictionary contains one of the allowlisted_paths, analysis should be carried out.

The attribute paths should overrule these lists.

This feature should be available for all detectors that may be analyzing all available parser matches, such as the VTD.

**Comment by dITySoftware** [comment]
The ECD already has something like allowlisted_paths - it is called paths there. In the VTD it is called path_list. Do these parameters need to be renamed to paths? Or is it enough the rename allowlisted_paths to blocklisted_paths in the ECD?

**Comment by landauermax** [comment]
I am not sure if its the same. As I understand, there should be for the ECD
1) paths: Allows the user to specify the exact paths of which the values should be analyzed for correlations
2) allowlisted_paths: Allows the user to specify branches of the parser tree for event correlation, e.g., the user specifies ['/parser/A', '/parser/B'] and the ECD will for example generate a correlation rule '/parser/A/x' -> '/parser/B/y', but will ignore for example '/parser/C/...'
3) blocklisted_paths: Allows the user to specify branches that should be ignored for event correlation, for example, specifying ['/parser/C'] would mean that only events that do not involve this path are considered.

**Comment by landauermax** [comment]
And similar structures should be available for all detectors that analyze all available paths.

**Comment by dITySoftware** [comment]
> I am not sure if its the same. As I understand, there should be for the ECD
> 
> 1. paths: Allows the user to specify the exact paths of which the values should be analyzed for correlations
> 2. allowlisted_paths: Allows the user to specify branches of the parser tree for event correlation, e.g., the user specifies ['/parser/A', '/parser/B'] and the ECD will for example generate a correlation rule '/parser/A/x' -> '/parser/B/y', but will ignore for example '/parser/C/...'
> 3. blocklisted_paths: Allows the user to specify branches that should be ignored for event correlation, for example, specifying ['/parser/C'] would mean that only events that do not involve this path are considered.

paths is same as allowlisted_paths in the ECD. If paths is not defined, all paths are analyzed. If it is defined, it works like you described allowlisted_paths. All other detectors already implement it that way.

Therefore this Issue only needs to refactor allowlisted_paths to blocklisted_paths in the ECD. Am I right or am I missing something?

**Comment by landauermax** [comment]
Only case 1. considers the values of the paths for correlation, case 2. and case 3. are for the correlation of events, therefore paths and allowedlist_paths cannot be the same.

If paths are defined, values of the paths are analyzed - allowedlist_paths and blocklist_paths are ignored in this case (and maybe should raise a warning, because they are not used in this setting). If paths are not defined, then allowedlist_paths and blocklist_paths are used to select/filter events that have or have not the paths as part of the event. In my example, the user only specifies the parent paths ['/parser/A', '/parser/B'] and then these paths and their children, e.g., /parser/A/x', are considered.

**Comment by dITySoftware** [comment]
> Only case 1. considers the values of the paths for correlation, case 2. and case 3. are for the correlation of events, therefore paths and allowedlist_paths cannot be the same.
> 
> If paths are defined, values of the paths are analyzed - allowedlist_paths and blocklist_paths are ignored in this case (and maybe should raise a warning, because they are not used in this setting). If paths are not defined, then allowedlist_paths and blocklist_paths are used to select/filter events that have or have not the paths as part of the event. In my example, the user only specifies the parent paths ['/parser/A', '/parser/B'] and then these paths and their children, e.g., /parser/A/x', are considered.

Ok I think I understand now. The issue is that blocklisted_paths are checked only if paths is None.  Would moving the blacklisted checks outside of the if paths is None block and changing it to
```
for allowlisted in self.allowlisted_paths:
                if allowlisted in parser_match.get_match_dictionary().keys():
                    return
```
solve the issue?

**Comment by landauermax** [comment]
Yes, I think so.

**Comment by landauermax** [comment]
I think the current PR280 only fixes a part of it. There should be both allowlisted_paths and blocklisted_paths, and either of them or both of them should be used in combination to select desired branches of the parser tree.

So if allowedlist_paths is set, only events with one of these paths are considered.
If blocklist_paths is set, only events that do not involve one of these paths are considered.

**Comment by dITySoftware** [comment]
The ECD already has the parameter 'paths' which does exactly that.

**Comment by landauermax** [comment]
I dont think so, since "paths" is just for the case where value (combinations) define the events. However, allowedlist_paths and blocklist_paths should be used when the event is defined by the full path list of the atoms.

**Comment by dITySoftware** [comment]
I have added allowlisted_paths now. These are only used when full paths are used (paths=None).
If the match does not contain at least one allowlisted path, it returns.

Please check the last commit.

**Comment by landauermax** [comment]
Thanks. The same functionality should also be added to the VTD and VCD, since they also analyze all tokens.

**Comment by dITySoftware** [comment]
> Thanks. The same functionality should also be added to the VTD and VCD, since they also analyze all tokens.

Okay i have added two new issues as the VCD cannot be published yet.

**Comment by dITySoftware** [comment]
I have implemented the functionality for VTD in #294

**Comment by landauermax** [comment]
Note that this is related to https://github.com/ait-aecid/logdata-anomaly-miner/issues/301
